File: TermiRustMobile/SSH/TmuxBootstrap.swift
import Foundation

struct TmuxBootstrap {
    let host: MobileHost

    func startupCommand() -> String? {
        guard host.persistentSession.enabled else {
            return host.startupCommand
        }

        let session = shellSingleQuote(host.persistentSession.sessionName ?? defaultSessionName())
        let attach = host.persistentSession.detachOthers
            ? "exec tmux attach-session -d -t \(session)"
            : "exec tmux attach-session -t \(session)"

        var create = "exec tmux new-session -s \(session)"
        if let startupDirectory = host.startupDirectory, !startupDirectory.isEmpty {
            create += " -c \(shellSingleQuote(startupDirectory))"
        }
        if let startupCommand = host.startupCommand, !startupCommand.isEmpty {
            create += " \(shellSingleQuote(startupCommand))"
        }

        return """
        if command -v tmux >/dev/null 2>&1; then
          if tmux has-session -t \(session) 2>/dev/null; then
            \(attach)
          else
            \(create)
          fi
        else
          printf 'TermiRust mobile persistent sessions require tmux on this host. Install tmux, then reconnect.\\n' >&2
          exec "${SHELL:-/bin/sh}"
        fi
        """
    }

    private func defaultSessionName() -> String {
        let raw = "tr-\(host.username)-\(host.host)-\(host.port)"
        let allowed = CharacterSet.alphanumerics.union(CharacterSet(charactersIn: "-"))
        return raw.unicodeScalars.map { allowed.contains($0) ? Character($0) : "-" }.map(String.init).joined()
    }
}

func shellSingleQuote(_ value: String) -> String {
    "'\(value.replacingOccurrences(of: "'", with: "'\"'\"'"))'"
}

File: apps/android/app/src/main/java/com/termirust/mobile/ssh/TmuxBootstrap.kt
package com.termirust.mobile.ssh

import com.termirust.mobile.data.MobileHost

class TmuxBootstrap(private val host: MobileHost) {
    fun startupCommand(): String? {
        if (!host.persistentSession.enabled) {
            return host.startupCommand
        }

        val session = shellSingleQuote(sessionName())
        val attach = if (host.persistentSession.detachOthers) {
            "exec tmux attach-session -d -t $session"
        } else {
            "exec tmux attach-session -t $session"
        }
        val create = buildString {
            append("exec tmux new-session -s $session")
            host.startupDirectory?.takeIf { it.isNotBlank() }?.let {
                append(" -c ${shellSingleQuote(it)}")
            }
            host.startupCommand?.takeIf { it.isNotBlank() }?.let {
                val shellCommand = "$it; exec \"${'$'}{SHELL:-/bin/sh}\" -l"
                append(" -- \"${'$'}{SHELL:-/bin/sh}\" -lc ${shellSingleQuote(shellCommand)}")
            }
        }

        return """
            if command -v tmux >/dev/null 2>&1; then
              if tmux has-session -t $session 2>/dev/null; then
                $attach
              else
                $create
              fi
            else
              printf 'TermiRust mobile persistent sessions require tmux on this host. Install tmux, then reconnect.\n' >&2
              exec "${'$'}{SHELL:-/bin/sh}"
            fi
        """.trimIndent()
    }

    internal fun sessionName(): String =
        host.persistentSession.sessionName?.takeIf { it.isNotBlank() } ?: defaultSessionName()

    private fun defaultSessionName(): String =
        "tr-${host.username}-${host.host}-${host.port}".replace(Regex("[^A-Za-z0-9-]"), "-")
}

fun shellSingleQuote(value: String): String = "'${value.replace("'", "'\"'\"'")}'"
